File: crates/calib/examples/planar_real.rs
//! Planar intrinsics calibration with real images.
//!
//! This example demonstrates the full calibration workflow with real data:
//! 1. Load calibration images from disk
//! 2. Detect chessboard corners
//! 3. Run initialization (Zhang's method with iterative distortion)
//! 4. Run non-linear optimization (bundle adjustment)
//! 5. Export and inspect results
//!
//! Run with: `cargo run -p calib --example planar_real`
//!
//! Dataset: Uses left camera images from `data/stereo/imgs/leftcamera/`

use anyhow::{Context, Result};
use calib::planar_intrinsics::{
    run_calibration_with_filtering, step_init, step_optimize, FilterOptions,
};
use calib::prelude::*;
use calib_targets::chessboard::ChessboardDetectionResult;
use calib_targets::{detect, ChessboardParams};
use chess_corners::ChessConfig;
use std::io::{self, Write};
use std::path::{Path, PathBuf};

// Board parameters (matches stereo_linear.json)
const BOARD_ROWS: u32 = 7;
const BOARD_COLS: u32 = 11;
const SQUARE_SIZE_M: f64 = 0.03; // 30mm squares

fn main() -> Result<()> {
    println!("=== Planar Intrinsics Calibration (Real Images) ===\n");

    let repo_root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../..");
    let imgs_dir = repo_root.join("data/stereo/imgs/leftcamera");

    if !imgs_dir.exists() {
        anyhow::bail!(
            "Dataset not found at {}\nPlease ensure the stereo dataset is available.",
            imgs_dir.display()
        );
    }

    println!("Dataset: {}", imgs_dir.display());
    println!(
        "Board: {}x{}, square size {:.1}mm\n",
        BOARD_ROWS,
        BOARD_COLS,
        SQUARE_SIZE_M * 1000.0
    );

    // Detect chessboard corners in all images
    println!("Detecting chessboard corners...");
    let views = load_views_with_progress(&imgs_dir)?;
    println!("\nLoaded {} views\n", views.len());

    if views.len() < 3 {
        anyhow::bail!("Need at least 3 views for calibration, got {}", views.len());
    }

    // Create dataset
    let dataset = PlanarDataset::new(views.into_iter().map(View::without_meta).collect())?;

    // Create calibration session
    let mut session = CalibrationSession::<PlanarIntrinsicsProblem>::new();
    session.set_input(dataset)?;

    // Step-by-step calibration
    println!("--- Step 1: Initialization ---");
    step_init(&mut session, None)?;

    let init_k = session.state.initial_intrinsics.as_ref().unwrap();
    let init_dist = session.state.initial_distortion.as_ref().unwrap();
    println!(
        "  Intrinsics: fx={:.1}, fy={:.1}, cx={:.1}, cy={:.1}",
        init_k.fx, init_k.fy, init_k.cx, init_k.cy
    );
    println!(
        "  Distortion: k1={:.4}, k2={:.4}, p1={:.5}, p2={:.5}\n",
        init_dist.k1, init_dist.k2, init_dist.p1, init_dist.p2
    );

    println!("--- Step 2: Optimization ---");
    step_optimize(&mut session, None)?;

    let state = &session.state;
    println!("  Final cost: {:.2e}", state.final_cost.unwrap());
    println!(
        "  Mean reprojection error: {:.4} px",
        state.mean_reproj_error.unwrap()
    );
    println!();

    // Export results
    let export = session.export()?;
    let final_k = export.params.intrinsics();
    let final_dist = export.params.distortion();

    println!("--- Final Results ---");
    println!("  Intrinsics:");
    println!("    fx = {:.2}", final_k.fx);
    println!("    fy = {:.2}", final_k.fy);
    println!("    cx = {:.2}", final_k.cx);
    println!("    cy = {:.2}", final_k.cy);
    println!("  Distortion:");
    println!("    k1 = {:.6}", final_dist.k1);
    println!("    k2 = {:.6}", final_dist.k2);
    println!("    k3 = {:.6}", final_dist.k3);
    println!("    p1 = {:.6}", final_dist.p1);
    println!("    p2 = {:.6}", final_dist.p2);
    println!(
        "  Mean reprojection error: {:.4} px\n",
        export.mean_reproj_error
    );

    // Alternative: run with filtering for outlier removal
    println!("--- Alternative: With Outlier Filtering ---");
    let views2 = load_views_with_progress(&imgs_dir)?;
    let dataset2 = PlanarDataset::new(views2.into_iter().map(View::without_meta).collect())?;

    let mut session2 = CalibrationSession::<PlanarIntrinsicsProblem>::new();
    session2.set_input(dataset2)?;

    run_calibration_with_filtering(
        &mut session2,
        FilterOptions {
            max_reproj_error: 1.0, // Stricter threshold
            min_points_per_view: 10,
            remove_sparse_views: true,
        },
    )?;

    let export2 = session2.export()?;
    println!(
        "  Mean reprojection error (filtered): {:.4} px",
        export2.mean_reproj_error
    );

    Ok(())
}

fn load_views_with_progress(imgs_dir: &Path) -> Result<Vec<CorrespondenceView>> {
    let chess_config = ChessConfig::default();
    let board_params = ChessboardParams {
        expected_rows: Some(BOARD_ROWS),
        expected_cols: Some(BOARD_COLS),
        ..ChessboardParams::default()
    };

    // Find all left camera images
    let mut indices: Vec<usize> = std::fs::read_dir(imgs_dir)?
        .filter_map(|e| e.ok())
        .filter_map(|e| {
            let name = e.file_name();
            let name = name.to_str()?;
            let name = name.strip_prefix("Im_L_")?.strip_suffix(".png")?;
            name.parse().ok()
        })
        .collect();
    indices.sort();

    let total = indices.len();
    let mut views = Vec::new();
    let mut skipped = 0;

    for (i, idx) in indices.iter().enumerate() {
        print!(
            "\r  Processing image {}/{} (Im_L_{}.png)",
            i + 1,
            total,
            idx
        );
        io::stdout().flush()?;

        let path = imgs_dir.join(format!("Im_L_{}.png", idx));
        match detect_chessboard(&path, &chess_config, board_params.clone()) {
            Ok(Some(view)) => views.push(view),
            Ok(None) => skipped += 1,
            Err(e) => {
                eprintln!("\nWarning: Failed to process Im_L_{}.png: {}", idx, e);
                skipped += 1;
            }
        }
    }

    println!(
        "\r  Processed {}/{} images ({} skipped)    ",
        total, total, skipped
    );

    Ok(views)
}

fn detect_chessboard(
    path: &Path,
    chess_config: &ChessConfig,
    board_params: ChessboardParams,
) -> Result<Option<CorrespondenceView>> {
    let img = image::ImageReader::open(path)
        .with_context(|| format!("Failed to open {}", path.display()))?
        .decode()
        .with_context(|| format!("Failed to decode {}", path.display()))?
        .to_luma8();

    let Some(detection) = detect::detect_chessboard(&img, chess_config, board_params) else {
        return Ok(None);
    };

    Ok(Some(detection_to_view(detection)?))
}

fn detection_to_view(detection: ChessboardDetectionResult) -> Result<CorrespondenceView> {
    let mut points_3d = Vec::new();
    let mut points_2d = Vec::new();

    for corner in detection.detection.corners {
        let Some(grid) = corner.grid else {
            continue;
        };
        points_3d.push(Pt3::new(
            grid.i as f64 * SQUARE_SIZE_M,
            grid.j as f64 * SQUARE_SIZE_M,
            0.0,
        ));
        points_2d.push(Pt2::new(corner.position.x as f64, corner.position.y as f64));
    }

    anyhow::ensure!(
        points_3d.len() >= 4,
        "Insufficient corners after grid filtering"
    );

    CorrespondenceView::new(points_3d, points_2d)
}

File: crates/vision-calibration-py/examples/planar_real.py
"""Planar intrinsics calibration from real stereo dataset images.

Python counterpart of:
`crates/vision-calibration/examples/planar_real.rs`
"""

from __future__ import annotations

import re
from pathlib import Path

import calib_targets
import numpy as np
from PIL import Image
import vision_calibration as vc

from _common import repo_root

BOARD_ROWS = 7
BOARD_COLS = 11
SQUARE_SIZE_M = 0.03


def load_gray(path: Path) -> np.ndarray:
    return np.asarray(Image.open(path).convert("L"), dtype=np.uint8)


def image_indices(folder: Path, prefix: str) -> list[int]:
    pattern = re.compile(rf"^{re.escape(prefix)}(\d+)\.png$")
    out: list[int] = []
    for path in folder.glob("*.png"):
        m = pattern.match(path.name)
        if m is not None:
            out.append(int(m.group(1)))
    out.sort()
    return out


def detect_chessboard_observation(path: Path) -> vc.Observation | None:
    image = load_gray(path)
    detection = calib_targets.detect_chessboard(
        image,
        chess_cfg=calib_targets.ChessConfig(),
        params=calib_targets.ChessboardParams(
            expected_rows=BOARD_ROWS,
            expected_cols=BOARD_COLS,
        ),
    )
    if detection is None:
        return None

    points_3d: list[tuple[float, float, float]] = []
    points_2d: list[tuple[float, float]] = []
    for corner in detection.detection.corners:
        if corner.grid is None:
            continue
        points_3d.append((float(corner.grid.i) * SQUARE_SIZE_M, float(corner.grid.j) * SQUARE_SIZE_M, 0.0))
        points_2d.append((float(corner.position[0]), float(corner.position[1])))
    if len(points_2d) < 4:
        return None
    return vc.Observation(points_3d=points_3d, points_2d=points_2d)


def load_dataset() -> vc.PlanarDataset:
    left_dir = repo_root() / "data/stereo/imgs/leftcamera"
    if not left_dir.exists():
        raise FileNotFoundError(f"dataset not found: {left_dir}")

    views: list[vc.PlanarView] = []
    skipped = 0
    for idx in image_indices(left_dir, "Im_L_"):
        obs = detect_chessboard_observation(left_dir / f"Im_L_{idx}.png")
        if obs is None:
            skipped += 1
            continue
        views.append(vc.PlanarView(observation=obs))

    if len(views) < 3:
        raise RuntimeError(f"need at least 3 usable views, got {len(views)}")
    print(f"Loaded {len(views)} views ({skipped} skipped)")
    return vc.PlanarDataset(views=views)


def main() -> None:
    print("=== Planar Intrinsics Calibration (Real Images, Python) ===")
    dataset = load_dataset()
    result = vc.run_planar_intrinsics(dataset)
    k = result.camera["k"]
    d = result.camera["dist"]

    print(f"Final cost: {result.final_cost:.3e}")
    print(f"Mean reprojection error: {result.mean_reproj_error:.6f} px")
    print(f"Intrinsics: fx={k['fx']:.3f}, fy={k['fy']:.3f}, cx={k['cx']:.3f}, cy={k['cy']:.3f}")
    print(f"Distortion: k1={d['k1']:.6f}, k2={d['k2']:.6f}, k3={d['k3']:.6f}, p1={d['p1']:.6f}, p2={d['p2']:.6f}")


if __name__ == "__main__":
    main()
